### PR TITLE
feat: index and fetch nft owners on a collection

### PR DESCRIPTION
**Pull request by jyeshe**
## What

- Index owners of a collection based on nft transfers  
- Add `/aex141/:contract_id/owners` endpoint for querying

## Why

Refs #629  

## Notes

- Includes handling delegated approvals
- An additional PR gonna provide docs for all aex141 endpoints
- New minting and burning gonna be handled by a separated PR (for transfer refactoring)

### CODE REVIEW

File: lib/ae_mdw/db/contract.ex [hunk 1]
```diff
@@ -288,6 +289,39 @@ defmodule AeMdw.Db.Contract do
...
+      |> State.put(Model.NftTokenOwner, m_token_owner)
+      |> State.delete(Model.NftOwnership, {old_owner_pk, contract_pk, token_id})
+      |> State.delete(Model.NftOwnerToken, {contract_pk, old_owner_pk, token_id})
+    else
+      error when error in [:not_found, false] ->
+        State.put(state2, Model.NftTokenOwner, m_token_owner)
+    end
+  end
+
+  def write_aex141_records(state, _pk, _txi, _i, _args), do: state
```
Comment (sborrazas) [inline lib/ae_mdw/db/contract.ex:323]: Isn't the `[from_pk, to_pk, token_id]` arguments type guaranteed by the fact that this is coming from an aex141 event hash?

File: lib/ae_mdw/db/contract.ex [hunk 2]
```diff
@@ -288,6 +289,39 @@ defmodule AeMdw.Db.Contract do
...
+    with {:ok, Model.nft_token_owner(index: {contract_pk, _token_id}, owner: old_owner_pk)} <-
+           State.get(state2, Model.NftTokenOwner, {contract_pk, token_id}),
+         true <- old_owner_pk != to_pk do
+      state2
+      |> State.put(Model.NftTokenOwner, m_token_owner)
+      |> State.delete(Model.NftOwnership, {old_owner_pk, contract_pk, token_id})
+      |> State.delete(Model.NftOwnerToken, {contract_pk, old_owner_pk, token_id})
+    else
+      error when error in [:not_found, false] ->
+        State.put(state2, Model.NftTokenOwner, m_token_owner)
```
Comment (sborrazas) [inline lib/ae_mdw/db/contract.ex:319]: We can move this line outside the `with` since it's always executed

File: lib/ae_mdw/db/contract.ex (hunk 1 above)
Comment (jyeshe) [reply lib/ae_mdw/db/contract.ex:323]: It's not guaranteed

File: lib/ae_mdw/db/contract.ex (hunk 2 above)
Comment (jyeshe) [reply lib/ae_mdw/db/contract.ex:319]: It needs to run after the `State.get(state2, Model.NftTokenOwner, {contract_pk, token_id})` but I can create a function to `maybe_delete_previous_owner`

Comment (sborrazas) [reply lib/ae_mdw/db/contract.ex:319]: Didn't realize that. It's fine as it is then!

**Review by sborrazas — APPROVED** [review]

**Review by thepiwo — APPROVED** [review]

File: lib/ae_mdw/db/contract.ex (hunk 2 above)
Comment (jyeshe) [reply lib/ae_mdw/db/contract.ex]: it is more readable, thanks

**Pull request merged**